src/cargo/sources/registry/mod.rs

```diff
@@ -709,26 +709,33 @@ impl<'cfg> Source for RegistrySource<'cfg> {
         kind: QueryKind,
         f: &mut dyn FnMut(Summary),
     ) -> Poll<CargoResult<()>> {
-        // If this is a precise dependency, then it came from a lock file and in
+        let mut req = dep.version_req().clone();
+
+        // Handle `cargo update --precise` here.
+        if let Some((_, requested)) = self
+            .source_id
+            .precise_registry_version(dep.package_name().as_str())
+            .filter(|(c, _)| req.matches(c))
+        {
+            req.lock_to(&requested);
+        }
+
+        // If this is a locked dependency, then it came from a lock file and in
         // theory the registry is known to contain this version. If, however, we
         // come back with no summaries, then our registry may need to be
         // updated, so we fall back to performing a lazy update.
-        if kind == QueryKind::Exact && dep.source_id().precise().is_some() && !self.ops.is_updated()
-        {
+        if kind == QueryKind::Exact && req.is_locked() && !self.ops.is_updated() {
             debug!("attempting query without update");
             let mut called = false;
-            ready!(self.index.query_inner(
-                dep.package_name(),
-                dep.version_req(),
-                &mut *self.ops,
-                &self.yanked_whitelist,
-                &mut |s| {
-                    if dep.matches(&s) {
+            ready!(self
+                .index
+                .query_inner(dep.package_name(), &req, &mut *self.ops, &mut |s| {
+                    if dep.matches(s.as_summary()) {
+                        // We are looking for a package from a lock file so we do not care about yank
```

Comment:
The to individually save changes above may mean that we can now `--precise` on a version that is locked. If that is a bug, and not a feature, we can filter out locked versions here as well. If it's a feature, we should add a test.

Reply:
I wanted to double check, is that typo and you mean "`--precise` on a version that is *yanked*"?

A related issue is #4225 and I feel like allowing people to select yanked versions (with a warning) is reasonable and aligns with #12425.

Note: I don't think this completely resolves #4225.

If we do make this change, we need to make sure its clearly communicated so we don't lose track of it at least for changelog generation.  In some cases like this, I've gone as far as having the refactor do extra work to maintain existing behavior and then had a separate commit that removed that just so its easier to discover that a change was made.

Also, if we do this and don't already have a test failing because of this, we should have a commit that adds a test showing the existing behavior and then the commit with the behavior change updates the test to show the new behavior.

Reply:
code updated so as to not add this interaction.